### PR TITLE
Added error codes

### LINKED ISSUE #9987: [suggestion] new ExchangeError: MaxPositionExceeded
For Binance this would be : `Error: -2027 - Exceeded the maximum allowable position at current leverage.`

### PR DESCRIPTION
**Pull request by ttodua**
As said in the following topic, fixes #9987

There are also more error codes missing from there, and should I fill others too? Also, will Futures type error codes conflict with Spot error codes?  Or they are still unique across whole binance alltogether?

### CODE REVIEW

File: js/binance.js [hunk 1]
```diff
@@ -915,6 +915,14 @@ module.exports = class binance extends Exchange {
                     '-2014': AuthenticationError, // { "code":-2014, "msg": "API-key format invalid." }
                     '-2015': AuthenticationError, // "Invalid API-key, IP, or permissions for action."
                     '-2019': InsufficientFunds, // {"code":-2019,"msg":"Margin is insufficient."}
+                    '-2020': ExchangeError, // {"code":-2020,"msg":"Unable to fill."}
+                    '-2021': InvalidOrder, // {"code":-2021,"msg":"Order would immediately trigger."}
+                    '-2022': InvalidOrder, // {"code":-2022,"msg":"ReduceOnly Order is rejected."}
+                    '-2023': InsufficientFunds, // {"code":-2023,"msg":"User in liquidation mode now."}
+                    '-2024': InsufficientFunds, // {"code":-2024,"msg":"Position is not sufficient."}
+                    '-2025': InsufficientFunds, // {"code":-2025,"msg":"Reach max open order limit."}
+                    '-2026': InsufficientFunds, // {"code":-2026,"msg":"This OrderType is not supported when reduceOnly."}
```
Comment (kroitor) [inline js/binance.js:924]: I would say, it's InvalidOrder here.

File: js/binance.js [hunk 2]
```diff
@@ -915,6 +915,14 @@ module.exports = class binance extends Exchange {
                     '-2014': AuthenticationError, // { "code":-2014, "msg": "API-key format invalid." }
                     '-2015': AuthenticationError, // "Invalid API-key, IP, or permissions for action."
                     '-2019': InsufficientFunds, // {"code":-2019,"msg":"Margin is insufficient."}
+                    '-2020': ExchangeError, // {"code":-2020,"msg":"Unable to fill."}
+                    '-2021': InvalidOrder, // {"code":-2021,"msg":"Order would immediately trigger."}
+                    '-2022': InvalidOrder, // {"code":-2022,"msg":"ReduceOnly Order is rejected."}
+                    '-2023': InsufficientFunds, // {"code":-2023,"msg":"User in liquidation mode now."}
+                    '-2024': InsufficientFunds, // {"code":-2024,"msg":"Position is not sufficient."}
+                    '-2025': InsufficientFunds, // {"code":-2025,"msg":"Reach max open order limit."}
```
Comment (kroitor) [inline js/binance.js:923]: I would set it to InvalidOrder here.

File: js/binance.js [hunk 3]
```diff
@@ -915,6 +915,14 @@ module.exports = class binance extends Exchange {
...
+                    '-2026': InsufficientFunds, // {"code":-2026,"msg":"This OrderType is not supported when reduceOnly."}
+                    '-2027': InsufficientFunds, // {"code":-2027,"msg":"Exceeded the maximum allowable position at current leverage."}
```
Comment (kroitor) [inline js/binance.js:925]: I think this one should also be InvalidOrder, cause we don't have a more specific exception for this kind of situation.

**Comment by ttodua** [comment]
@kroitor I am following these two docs:
https://binance-docs.github.io/apidocs/spot/en/
https://binance-docs.github.io/apidocs/delivery/en/ ( delivery & futures had the same errors, I've checked all of them)
till 4xxx they had mostly similar errors,but after 4xxx I have to do differently - to collect the 'conflicting' ones and add them in deepExtend in 'futures/delivery' exchange class separately, to override spot's 4xxx ones (as they are different)

**Comment by kroitor** [comment]
@ttodua we will still have to make the base binance class aware of them, cause most users will not use the shorthand aliases like binanceusdm and binancecoinm, and will instead use binance with defaultType future or with defaultType delivery.

**Comment by ttodua** [comment]
@kroitor do you remind an example of other exchange class, where I can see how the two-same error codes are being handled like that way (according to 'defaultType' or alike?) Is there any currently available structure for that?

**Comment by kroitor** [comment]
@ttodua check these lines: https://app.travis-ci.com/github/ccxt/ccxt/builds/244094314#L1185

**Comment by kroitor** [comment]
> @kroitor do you remind an example of other exchange class, where I can see how the two-same error codes are being handled like that way (according to 'defaultType' or alike?) Is there any currently available structure for that?

No, i don't think we have a specific handler, cause usually those codes don't intersect. I would suggest to collect the exceptions that we want to add first, and then structure them accordingly depending on the exceptions themselves.

File: js/binance.js (hunk 3 above)
Comment (ttodua) [reply js/binance.js]: Ah, I've not checked those ones at all, had copy-pasted that initially, my fault. 
revised and fixed them.

**Comment by ttodua** [comment]
> @ttodua check these lines: https://app.travis-ci.com/github/ccxt/ccxt/builds/244094314#L1185

yeah, I was in between changes, missed it. Atm finished revision and completing for spot. 
Will post a note in below comment I have obserbed.

**Comment by ttodua** [comment]
1) I've also removed JSON-formatted objects from comment, instead left only MESSAGE's themselves
2) After I checked all of them against each other, the differences I've found in docs among Spot/Futures/Options:

Futures (usdm & coinMargined) and Options have mostly same error-codes up to Fourth-Thousands (4xxx), except the below ones:
- 1004 > different: futures vs options
- 1004  > different: futures vs spot
- 1004  > different: spot vs options
- 1013  > different: futures vs spot
- 1128  > different: futures vs options
- 2027  > different: futures vs options
- 2027  > different: options vs spot
- 5005 > different: options vs spot
- 5008  > different: options vs spot

Fourth-thousands row (`4xxxx`) are different among all of types (spot/futures/coinm/usdm/options)

3) there is a bug in docs: https://binance-docs.github.io/apidocs/spot/en/#5012-transfer_pending 5012 is doubled on place of (possibly) 5022 . Please report that to binance if you want, so they corrected (idk if that's 5022 exactly or 5032 or what)
4) 9xxx series are nonstandard : https://binance-docs.github.io/apidocs/spot/en/#9xxx-filter-failures

**Comment by kroitor** [comment]
> 1. I've also removed JSON-formatted objects from comment, instead left only MESSAGE's themselves

Tbh, i'd keep them as they were, if possible, please revert that change.

> 2. After I checked all of them against each other, the differences I've found in docs among Spot/Futures/Options:

Let's review all the differences.
Take, say, code 1004 – what is different about it? How does the error message look like with different defaultTypes?

> 3. there is a bug in docs: https://binance-docs.github.io/apidocs/spot/en/#5012-transfer_pending 5012 is doubled on place of (possibly) 5022 . Please report that to binance if you want, so they corrected (idk if that's 5022 exactly or 5032 or what)

Thanks, will retransmit it to them.

> 4. 9xxx series are nonstandard : https://binance-docs.github.io/apidocs/spot/en/#9xxx-filter-failures

Please, elaborate a bit on the nonstandard part.

Also, please, check the build status of this PR: https://github.com/ccxt/ccxt/pull/11068/checks?check_run_id=4661485318

**Comment by ttodua** [comment]
### 1) About specific differences ###

* 1004
[spot & options] SERVER_BUSY  : "Server is busy, please wait and try again"
[usdm & coinm] DUPLICATE_IP  : "This IP is already on the white list"

* 1013
[spot] When canceling the Rest API can now return errorCode -1013 OR -2011 if the symbol's status isn't TRADING.
[usdm & coinm] INVALID_MESSAGE  : "INVALID_MESSAGE"

* 1128
[spot & usdm & coinm] OPTIONAL_PARAMS_BAD_COMBO  : "Combination of optional parameters invalid."
[options] BAD_CONTRACT  : "Invalid underlying"

* 2027
[usdm & coinm] MAX_LEVERAGE_RATIO: "Exceeded the maximum allowable position at current leverage."
[options] OPTION_MARGIN_NOT_SUFFICIENT : "Option margin is insufficient."

* 5005
[spot]  USER_ASSET_AMOUNT_IS_TOO_LOW  :"The residual balances of %s is too low, Please re-choose."
[options]  BLACK_COUNTRY :"This is a blacklist country."

* 5008
[spot] ASSET_WITHDRAW_WITHDRAWING_NOT_ENOUGH : "Insufficient amount of returnable assets."
[options] KYC_NOT_PASS : "User kyc not pass."

(and in `4xxx` class all of those floors have conflicts)





### 2) About `9xxx` class ###
As linked https://binance-docs.github.io/apidocs/spot/en/#9xxx-filter-failures - I don't have experience how to handle these variative , non-assigned error-codes. I will need your guidance, or whoever already has done smth like this.


### 3) About {} json format inline comments ###
Asked in chat.

### 4) Build status ###
I was doing this patch/pr from github UI, so haven't run 'build' on my local. did now and it runs fine. I've removed extra space and now it might be fixed. If not, I'll need your help to fix the build issue.

**Comment by ttodua** [comment]
I think the above listed things are remaining into 'todo' status. (I'll open an issue for that and will mark with TODO label).
However, for now, I think I will need to switch to exchange-integration task you mentioned, and if this current PR is ok for integration, then after merging, I will create another PR to address the above-listed things after I will have a guidance from you with those "yet-unfamiliar-for-me"  parts.

**Pull request merged**